integrations/SD_Lon/calculate_primary.py

```diff
@@ -0,0 +1,321 @@
+            dates.add(datetime.datetime.strptime(eng['validity']['from'],
+                                                 '%Y-%m-%d'))
+            if eng['validity']['to']:
+                to = datetime.datetime.strptime(eng['validity']['to'], '%Y-%m-%d')
+                day_after = to + datetime.timedelta(days=1)
+                dates.add(day_after)
+            else:
+                dates.add(datetime.datetime(9999, 12, 30, 0, 0))
+
+        date_list = sorted(list(dates))
+        logger.debug('List of cut-dates: {}'.format(date_list))
+        return date_list
+
+    def _read_engagement(self, date):
+        mo_engagement = self.helper.read_user_engagement(
+            user=self.mo_person['uuid'],
+            at=date,
+            only_primary=True,  # Do not read extended info from MO.
+            use_cache=False
+        )
+        return mo_engagement
+
+    def check_all_for_primary(self):
+        """
+        Check all users for the existence of primary engagements.
+        :return: TODO
+        """
+        # TODO: This is a seperate function in AD Sync! Change to mora_helpers!
+        count = 0
+        all_users = self.helper.read_all_users()
+        for user in all_users:
+            if count % 250 == 0:
+                print('{}/{}'.format(count, len(all_users)))
+            count += 1
+
+            self.set_current_person(uuid=user['uuid'])
+            date_list = self._find_cut_dates()
+            for i in range(0, len(date_list) - 1):
+                date = date_list[i]
+                mo_engagement = self._read_engagement(date)
+                primary_count = 0
+                for eng in mo_engagement:
+                    if eng['engagement_type']['uuid'] in self.primary:
+                        primary_count += 1
+            if primary_count == 0:
+                print('No primary for {} at {}'.format(user['uuid'], date))
+            elif primary_count > 1:
+                # This will typically happen because of both a primary and a status0
+                logger.info('{} has more than one primary'.format(user['uuid']))
+                extra_primary_count = 0
+                for eng in mo_engagement:
+                    if eng['engagement_type']['uuid'] in self.primary[0:2]:
+                        extra_primary_count += 1
+                if extra_primary_count == 1:
+                    logger.info('Only one primary was different from status 0')
+                if extra_primary_count > 1:
+                    print('Too many primaries for {} at {}'.format(user['uuid'],
+                                                                   date))
+            else:
+                # print('Correct')
+                pass
+
+    def recalculate_primary(self):
+        """
+        Re-calculate primary engagement for the enire history of the current user.
+        """
+        logger.info('Calculate primary engagement: {}'.format(self.mo_person))
+        date_list = self._find_cut_dates()
```

Comment:
Du læser alle engagementer for brugeren her?

Reply:
Ja.